File: 103.binary-tree-zigzag-level-order-traversal.py
#
# @lc app=leetcode id=103 lang=python3
#
# [103] Binary Tree Zigzag Level Order Traversal
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        return [
            level if idx % 2 == 0 else list(reversed(level))
            for idx, level in enumerate(self.visitNodes([root]))
        ]

    def visitNodes(self, nodes) -> List[List[int]]:
        nextLevel = []
        for n in nodes:
            if n.left:
                nextLevel.append(n.left)
            if n.right:
                nextLevel.append(n.right)
        # base case
        if len(nextLevel) == 0:
            return [[n.val for n in nodes]]
        return [[n.val for n in nodes]] + self.visitNodes(nextLevel)


# @lc code=end

File: 103.binary-tree-zigzag-level-order-traversal.go
/*
 * @lc app=leetcode id=103 lang=golang
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
package main

/*
define a recursive function: visitLevel(nodes []*TreeNode, leftToRight bool)
if leftToRight is true, this level should be traverse left to right, else right to left

visitLevel(nodes, leftToRight){

 1. traverse this level in the order specified by the 2nd parameter

 2. return [thisLevelTraversalOutput, visitLevel(nodesAtNextLevel)...]

    base case, if curr level, no nodes, return []
    }

return visitLevel([root], true)
*/
func zigzagLevelOrder(root *TreeNode) [][]int {

	var visitLevel func(nodes []*TreeNode, l2r bool) [][]int
	visitLevel = func(nodes []*TreeNode, l2r bool) [][]int {
		if len(nodes) == 0 {
			return [][]int{}
		}
		currLevel := make([]int, len(nodes))
		if l2r == true {
			for i, n := range nodes {
				currLevel[i] = n.Val
			}
		} else {
			for i, n := range nodes {
				currLevel[len(nodes)-1-i] = n.Val
			}
		}
		nextLevel := []*TreeNode{}
		for _, n := range nodes {
			if n.Left != nil {
				nextLevel = append(nextLevel, n.Left)
			}
			if n.Right != nil {
				nextLevel = append(nextLevel, n.Right)
			}
		}
		ans := [][]int{
			currLevel,
		}
		ans = append(ans, visitLevel(nextLevel, !l2r)...)
		return ans
	}
	if root == nil {
		return [][]int{}
	}
	return visitLevel([]*TreeNode{root}, true)
}

// @lc code=end
